crates/storage/src/collections/stash/mod.rs

```diff
@@ -485,6 +485,34 @@ where
         }
     }
 
+    /// Drains this stash, calls `f` with each drained `T`.
+    pub fn drain_with<F>(&mut self, mut f: F)
```

Comment:
Maybe similar to the `defrag` methods we could use some `Option<usize>` parameter that when `Some(value)` limits the maximum amount of elements that are going to be drained and when `None` simply has no limit on the number of drained elements.

Reply:
Another option is to turn `F` into `FnMut(T) -> bool` where the return value `bool` states if more elements shall be drained if possible. So `true` means: Please drain more elements whereas `false` means, please stop draining more elements. This would be even more flexible.

Reply:
So it's a good idea and I added the latter version.

Imo `HashMap::drop` should always `drop` everything, that's the expectation. There should not be an upper bound where it stops. Hence I still `drain_with` everything. A user can still use `drain_with` with some user-defined upper bound and appropriate handling if it's violated. So the user has some way of controlling sprawling gas costs.